SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialEventTracer.h"
+
+#include "Misc/ScopeLock.h"
+#include "SpatialGDKSettings.h"
+#include "UObject/Object.h"
+#include "UObject/UnrealType.h"
+#include <WorkerSDK/improbable/c_io.h>
+#include <WorkerSDK/improbable/c_trace.h>
+
+DEFINE_LOG_CATEGORY(LogSpatialEventTracer);
+
+using namespace SpatialGDK;
+using namespace worker::c;
+
+SpatialEventTracerGuard EventTracerGuard;
+
+void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+{
+	FScopeLock ScopeLock(&EventTracerGuard.CriticalSection);
+
+	SpatialEventTracer* EventTracer = EventTracerGuard.EventTracer;
+	if (EventTracer == nullptr || !EventTracer->IsEnabled())
+	{
+		return;
+	}
+
+	if (!ensure(EventTracer->Stream.Get() != nullptr))
+	{
+		return;
+	}
+
+	uint32_t ItemSize = Trace_GetSerializedItemSize(Item);
+	if (EventTracer->BytesWrittenToStream + ItemSize <= EventTracer->MaxFileSize)
```

Comment:
Should we close the stream once this condition is false? If there are 10 bytes remaining and 11 bytes is attempted followed by 9 we would omit a message

Reply:
Just wanted to understand how 11 bytes would be attempted if there was only 10 bytes left?

Reply:
Maybe just setting BytesWrittenToStream = EventTracer->MaxFileSize will do

Reply:
Confused, what are we trying to avoid here?

Reply:
If `EventTracer->BytesWrittenToStream` is 10 bytes smaller than `EventTracer->MaxFileSize`, and the current `ItemSize` is 11 bytes, we'll skip it. If the following item is 10 bytes or smaller, we will write it, thus skipping an item.

Reply:
Ahh yeah, nice.